src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,450 @@
+
+    std::vector<seastar::future<std::tuple<Status, Payload> > > readFutures;
+    for (auto& log_entry: metadataInfo){
+        readFutures.push_back(_client.read(log_entry.plogId, log_entry.start_offset, log_entry.size));
+    }
+
+    return seastar::when_all_succeed(readFutures.begin(), readFutures.end())
+    .then([this] (std::vector<std::tuple<Status, Payload> >&& responses){
+        std::vector<Payload> outputPayloads;
+        for (auto& response: responses){
+            auto& [status, resp] = response;
+            if (!status.is2xxOK()){
+                throw k2::dto::PlogReadError("unable to read a plog");
+            }
+            outputPayloads.push_back(std::move(resp));
+        }
+        return seastar::make_ready_future<std::vector<Payload> >(std::move(outputPayloads));
+    });
+}
+
+
+seastar::future<Status> 
+LogStreamBase::reload(std::vector<dto::MetadataRecord> records){
+    K2LOG_I(log::lgbase, "LogStreamBase Reload");
+    if (_create){
+        throw k2::dto::LogStreamBaseExistError("Log stream already created");
+    }
+    _switched = false;
+    for (auto& record: records){
+        _usedPlogIdVector.push_back(record.plogId);
+        PlogInfo info{.currentOffset=record.sealed_offset, .sealed=true, .index=(uint32_t)_usedPlogIdVector.size()-1};
+        _usedPlogInfo[std::move(record.plogId)] = std::move(info);
+    }
+    _usedPlogInfo[_usedPlogIdVector[_usedPlogIdVector.size()-1]].sealed=false;
+
+    std::vector<seastar::future<std::tuple<Status, String> > > createFutures;
+    for (uint32_t i = 0; i < PLOG_POOL_SIZE; ++i){
+        createFutures.push_back(_client.create());
+    }
+    return seastar::when_all_succeed(createFutures.begin(), createFutures.end())
+    .then([this] (auto&& responses){
+        // persist the plog info 
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw k2::dto::PlogCreateError("unable to create plog for Logstream Base");
+            }
+            _redundantPlogPool.push_back(std::move(plogId));
+        }
+        _create = true;
+        return seastar::make_ready_future<Status>(Statuses::S200_OK("successfully reload metadata"));
+    });
+}
+
+seastar::future<std::tuple<Status, std::tuple<uint32_t, bool>>> 
+LogStreamBase::get_plog_status(String plogId){
+    return _client.getPlogStatus(std::move(plogId));
+}
+LogStream::LogStream() {
+    K2LOG_I(log::logstream, "dtor");
+}
+
+LogStream::~LogStream() {
+    K2LOG_I(log::logstream, "dtor");
+}
+
+seastar::future<> 
+LogStream::init(String name, MetadataMgr* metadataMgr, String cpo_url, String persistenceClusrerName){
+    _name = name;
+    _metadataMgr = metadataMgr;
+    return init_plog_client(cpo_url, persistenceClusrerName);
+}
+
+seastar::future<Status>
+LogStream::_persistSelfMetadata(uint32_t sealed_offset, String new_plogId){
+    return _metadataMgr->persistMetadata(_name, sealed_offset, std::move(new_plogId));
```

Comment:
AddNewPLogIntoLogStream(_logStreamName/Id, previousPlog_sealed_offset, new_plogId)

AddNewPlog(previousPlog_sealed_offset, new_plogId) // for self

Reply:
changed